experimental/virtio/src/console.rs

```diff
@@ -0,0 +1,215 @@
+    pub fn read_bytes(&mut self) -> Option<VecDeque<u8>> {
+        let buffer = self.rx_queue.read_next_used_buffer()?;
+        let mut result = VecDeque::with_capacity(buffer.len());
+        result.extend(buffer);
+
+        if self.rx_queue.inner.must_notify_device() {
+            // Notify the device that bytes have been read and the buffer is available again.
+            self.device.notify_queue(RX_QUEUE_ID);
+        }
+
+        Some(result)
+    }
+
+    /// Writes the data to the transmit queue.
+    ///
+    /// Returns the number of bytes written, if any.
+    pub fn write_bytes(&mut self, data: &[u8]) -> Option<usize> {
+        let result = self.tx_queue.write_buffer(data);
+
+        if self.tx_queue.inner.must_notify_device() {
+            // Notify the device that new bytes have been written.
+            self.device.notify_queue(TX_QUEUE_ID);
+        }
+
+        result
+    }
+
+    /// Gets the device status.
+    ///
+    /// See <https://docs.oasis-open.org/virtio/virtio/v1.1/csprd01/virtio-v1.1-csprd01.html#x1-100001>.
+    pub fn get_status(&self) -> u32 {
+        self.device.get_status()
+    }
+
+    fn new(device: VirtioBaseDevice<VirtioPciTransport>) -> Self {
+        let tx_queue = DriverWriteOnlyQueue::new();
+        let rx_queue = DeviceWriteOnlyQueue::new();
+        Console {
+            device,
+            tx_queue,
+            rx_queue,
+            pending_data: None,
+        }
+    }
+
+    /// Initializes the device and configures the queues.
+    fn init(&mut self) -> anyhow::Result<()> {
+        self.device
+            .start_init(DEVICE_ID as u32)
+            .map_err(|error| anyhow::anyhow!("Virtio error: {:?}", error))
+            .context("Couldn't initialize the PCI device.")?;
+        self.device
+            .configure_queue(
+                RX_QUEUE_ID,
+                QUEUE_SIZE as u16,
+                self.rx_queue.inner.get_desc_addr(),
+                self.rx_queue.inner.get_avail_addr(),
+                self.rx_queue.inner.get_used_addr(),
+            )
+            .map_err(|error| anyhow::anyhow!("Queue configuration error: {:?}", error))
+            .context("Couldn't configure the receive queue.")?;
+        self.device
+            .configure_queue(
+                TX_QUEUE_ID,
+                QUEUE_SIZE as u16,
+                self.tx_queue.inner.get_desc_addr(),
+                self.tx_queue.inner.get_avail_addr(),
+                self.tx_queue.inner.get_used_addr(),
+            )
+            .map_err(|error| anyhow::anyhow!("Queue configuration error: {:?}", error))
+            .context("Couldn't configure the transmit queue.")?;
+        self.device
+            .complete_init()
+            .map_err(|error| anyhow::anyhow!("Device activation error: {:?}", error))
+            .context("Couldn't activate the PCI device.")?;
+        Ok(())
+    }
+
+    /// Tries once to fill the destination with as much data as is currently available, either in
```

Comment:
Is there a reason we try the pending buffer first?

Reply:
Great, maybe we want to make it explicit that we first want to read the pending buffer, i.e., "first in the pending buffer" and "then from the next available buffer". 

Just a short clarification: If we cannot fully read `pending_data`, we move the remaining `pending_data` back to `pending_data`.

Reply:
Added more clarification to the comments. I think "first in the pending buffer" and "then from the next available buffer" could be misleading. The "either/or" nature is important.